### PR TITLE
BUG: df.__setitem__ casting ea to object when setting with scalar key and  frame value

### LINKED ISSUE #20476: Assigning with partial indexing on columns and mixed type unnecessarily upcasts
#### Code Sample, a copy-pastable example if possible

```python
In [2]: df = pd.DataFrame(-1, index=range(3), columns=pd.MultiIndex.from_product([['A', 'B'], ['a', 'b', 'c']]))

In [3]: filler = pd.DataFrame([[1, 2, 3.]]*3, index=range(3), columns=['a', 'b', 'c'])

In [4]: filler
Out[4]: 
   a  b    c
0  1  2  3.0
1  1  2  3.0
2  1  2  3.0

In [5]: df['A'] = filler

In [6]: df
Out[6]: 
     A            B      
     a    b    c  a  b  c
0  1.0  2.0  3.0 -1 -1 -1
1  1.0  2.0  3.0 -1 -1 -1
2  1.0  2.0  3.0 -1 -1 -1
```

#### Problem description

Only one of the three columns had float dtype, but the two others have been upcasted.

My guess is that somewhere a ``.values`` is used inappropriately.

#### Expected Output

``` python
In [6]: df
Out[6]: 
   A          B      
   a  b    c  a  b  c
0  1  2  3.0 -1 -1 -1
1  1  2  3.0 -1 -1 -1
2  1  2  3.0 -1 -1 -1
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-5-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.23.0.dev0+653.g7273ea070
pytest: 3.2.3
pip: 9.0.1
setuptools: 38.5.2
Cython: 0.25.2
numpy: 1.14.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### LINKED ISSUE #46896: BUG: incorrect type when casting to nullable type in multiindex dataframe
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame(
    columns=pd.MultiIndex.from_tuples([("a", "c"), ("a", "d")]),
    data=[[1, 2], [3, 4]],
)

df["a"] = df["a"].astype("Int64")

print(df.dtypes)  # results in type "object" for both columns

print(df["a"].astype("Int64").dtypes)  # here the dtypes are correct
```

### Issue Description

When casting to a nullable type (like "Int64") in a multiindex column dataframe, the resulting type is "object" (where it should be "Int64").

If you cast to a numpy type (e.g. `.astype(float)`), we do see the expected behavior (namely, a dataframe with types `float`). 

It seems that the problem is not with the `.astype(..)` method, but with the `__setitem__` (i.e., with the re-assignment `df["a"] = df["a"]...`).

Happy to help, but it seems it requires debugging C code which is beyond the capabilities of my tiny brain. 

### Expected Behavior

We expect that the columns have the right type. 

### Installed Versions

INSTALLED VERSIONS
------------------
commit           : 9f2491875945d4b515f66491dbe507632060d6bf
python           : 3.8.13.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.4.0
Version          : Darwin Kernel Version 21.4.0: Fri Mar 18 00:47:26 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T8101
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : None.UTF-8

pandas           : 1.5.0.dev0+725.g9f24918759.dirty
numpy            : 1.21.6
pytz             : 2022.1
dateutil         : 2.8.2
pip              : 22.0.4
setuptools       : 62.1.0
Cython           : 0.29.28
pytest           : 7.1.2
hypothesis       : 6.45.1
sphinx           : 4.5.0
blosc            : None
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.8.0
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.1.1
IPython          : 8.3.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : 1.3.4
brotli           : 
fastparquet      : 0.8.0
fsspec           : 2021.11.0
gcsfs            : 2021.11.0
markupsafe       : 2.1.1
matplotlib       : 3.5.1
numba            : 0.55.1
numexpr          : 2.8.0
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : 1.1.5
pyxlsb           : None
s3fs             : 2021.11.0
scipy            : 1.8.0
snappy           : 
sqlalchemy       : 1.4.36
tables           : 3.7.0
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
zstandard        : None

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #46896 (Replace xxxx with the Github issue number)
- [x] closes #20476 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Setting with duplicate columns is still buggy, but could not figure this out efficiently yet, will open a follow up when this is merged

### CODE REVIEW

File: pandas/tests/frame/indexing/test_coercion.py [hunk 1]
```diff
@@ -164,7 +164,6 @@ def test_12499():
     tm.assert_frame_equal(df, expected)
 
 
-@pytest.mark.xfail(reason="Too many columns cast to float64")
```
Comment (mroeschke) [inline pandas/tests/frame/indexing/test_coercion.py:167]: So this essentially tests https://github.com/pandas-dev/pandas/issues/46896?

Comment (phofl) [reply pandas/tests/frame/indexing/test_coercion.py:167]: Yep correct

Comment (phofl) [reply pandas/tests/frame/indexing/test_coercion.py:167]: It's basically a duplicate of #20476

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4064,13 +4064,27 @@ def _set_item_frame_value(self, key, value: DataFrame) -> None:
...
+                    value = value.reindex(cols_droplevel, axis=1)
 
-        # now align rows
-        arraylike = _reindex_for_setitem(value, self.index)
-        self._set_item_mgr(key, arraylike)
+                for col, col_droplevel in zip(cols, cols_droplevel):
+                    self[col] = value[col_droplevel]
+                return
+
+            if is_scalar(cols):
```
Comment (mroeschke) [inline pandas/core/frame.py:4075]: Which test hits this case?

Comment (phofl) [reply pandas/core/frame.py:4075]: test_setitem_frame_overwrite_with_ea_dtype

This is the case where we overwrite an existing column

File: pandas/core/frame.py [hunk 3]
```diff
@@ -4064,13 +4064,27 @@ def _set_item_frame_value(self, key, value: DataFrame) -> None:
...
+                self[cols] = value[value.columns[0]]
+                return
+
+            # now align rows
+            arraylike = _reindex_for_setitem(value, self.index)
+            self._set_item_mgr(key, arraylike)
+            return
+
+        if len(value.columns) != 1:
+            raise ValueError("Expected DataFrame with 1 column.")
```
Comment (mroeschke) [inline pandas/core/frame.py:4085]: Maybe `f"Cannot set a DataFrame with multiple columns to the single column {key}"` is a bit more specific?

Comment (phofl) [reply pandas/core/frame.py]: Works for me, changed

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**